### ISSUE TITLE
AmbigousLookup with Union of Literals (should resolve correctly)

### ISSUE THREAD

**Issue by PhilipVinc** [beartype/plum#99]
The following MWE shows a case where a more Specific signature (the second) should have precedence over the former, but it has not and leads to an ambiguity error.

I'm not sure what is causing it in here.

```python
from plum import dispatch

from typing import Literal, Union, Any


@dispatch
def f(x: Union[Literal[True], Literal[False]], b: Any):
    return 1

@dispatch
def f(x: Literal[True], b: int):
    return 2

f(False, 1) # works, returns 1

f(True, 'a') # works, returns 1

f(True, 1) # fails


AmbiguousLookupError: For function `f`, `(True, 1)` is ambiguous among the following:
  Signature(typing.Union[typing.Literal[True], typing.Literal[False]], typing.Any, implementation=<function f at 0x105300ee0>) (precedence: 0)
  Signature(typing.Literal[True], int, implementation=<function f at 0x105703a60>) (precedence: 0)
```

**Comment by PhilipVinc** [comment]
reduced it down to a possibly bear type bug:

```python
In [1]: import beartype

In [2]: from beartype import door

In [3]: import typing

In [5]: beartype.door.TypeHint(typing.Literal[True])<=beartype.door.TypeHint(typing.Union[typing.Literal[True], typing.Literal[False]])
Out[5]: False

In [6]: beartype.door.TypeHint(typing.Literal[True])>=beartype.door.TypeHint(typing.Union[typing.Literal[True], typing.Literal[False]])
Out[6]: False
```

where clearly one of the two should be True.

**Comment by wesselb** [comment]
Hey @PhilipVinc! Ah, very good catch! I agree with your assessment that the first inequality should be true. Would you be able to open an issue on @beartype?

**Comment by PhilipVinc** [comment]
done already. Let's hope our bearish friend has a solution for that one

**Comment by wesselb** [comment]
Ah, fantastic! I should’ve checked first. :)

I’m certain that @leycec will know what’s up!

**Comment by leycec** [comment]
`beartype.door` is full to the brim with bugs! I cry live on GitHub.  :sob:

@leycec is lazy and can barely maintain the main @beartype codebase itself, let alone both @beartype *and* `beartype.door`. DOOR is a motile appendage bolted on to the side of @beartype like a creepy baby claw extending from the belly button. Please, somebody who is not me: help me maintain the shambolic object-oriented DOOR API that is now walking around with its arms outstretched begging passerby for "...brains! Moar brains! And tasty working PRs too, please."

**Comment by PhilipVinc** [comment]
@leycec thanks for the insight! 
@wesselb , this means that replacing `Val` with `Literal` is not entirely possible right now. Can we remove the deprecation warning for the time being until this is fixed in bear type itself?

**Comment by wesselb** [comment]
Sorry both for the delayed reply. I've just come back from two weeks of exploring the beauties of Corsica.

@PhilipVinc, definitely! Your PR has already been merged, so that should be good now. :)

@leycec, maintaing `beartype` is an incredibly tall ask because it's such a complex library, and we all more than grealty appreciate the time and effort you're putting into it. :) It's not more than reasonable that we try to help out wherever we can. Once I've caught up with everything again, I can try to figure out what's wrong and put together a PR if I manage.

**Comment by leycec** [comment]
> two weeks of exploring the beauties of Corsica.

![wow](https://media.tenor.com/y73I-JcKQi0AAAAC/corse-corsica.gif)

:open_mouth: 

> ...I can try to figure out what's wrong and put together a PR if I manage.

I'd be delighted to walk you through this. Delighted, I say! When copious free time erupts in your life again, here's exactly where your ~~agonizing gauntlet of pain~~ wondrous journey of self-discovery begins:

1. **Step 1.** Let's add a new use case exercising this issue to @beartype's test suite. Specifically:

```python
# In the "beartype_test/a00_unit/a40_api/door/_doorfixture.py" submodule:
def door_cases_subhint() -> 'Iterable[Tuple[object, object, bool]]':
    ...
    HINT_SUBHINT_CASES = [
        ...
        # ..................{ PEP 586                        }..................
        # PEP 586-compliant type hints.

        #FIXME: Add new 3-tuples resembling something like this!
        #@beartype: it is fun or educational..... possibly both.
        #These 3-tuples test that the following comparisons evaluate to the expected booleans:
        #    TypeHint(Literal[True]) <= TypeHint(Union[Literal[True], Literal[False]]) is True
        #    TypeHint(Union[Literal[True], Literal[False]]) <= TypeHint(Literal[True]) is False
        (Literal[True], Union[Literal[True], Literal[False]], True)
        (Union[Literal[True], Literal[False]], Literal[True], False)
        ...
     ]
     ...
```

2. **Step 2.** Refactor some stuff in the `beartype.door._cls.pep.doorpep586` submodule. So as not to burden anyone with too many lies **[**read: _I am tired and hungry_**]**, let's pretend we know what needs to be done here. It's probably not *too* arduous, because DOOR is a reasonably designed object-oriented API – unlike the rest of @beartype.
3. **Celebrate!**

:partying_face:

**Comment by leycec** [comment]
@wesselb: Actually... **belay that bad advice.** I couldn't help myself. I looked into this a bit tonight. It's *vastly* more gruesome than I thought. Because I am a glutton for punishment, I'm tackling this head-on. With any luck (and some sweaty foreheads and much bleeding profusely), I'll resolve this in time for @beartype's upcoming 0.16.0 release. *Serenity now!*

**Comment by leycec** [comment]
**Resolved by beartype/beartype@f604f05.** Of course, this also resolves beartype/beartype#271. Of course, I am now even more bald thanks to GitHub-related stress injuries incurred during this resolution. I cannot be held responsible if I now begin malding...

Oh, noes. It's happening. I'm malding! *I'M MAAAAAALDING!* :bald_man: `</awkward_cough>`

**Comment by wesselb** [comment]
Whoa, @leycec, thank you so much for tackling this head-on!! I see that you've produced an awesome post detailing exactly what went wrong and how you've fixed it, which I'm going to enjoy reading right now. :) I just wanted to express my gratitude once more: thank you! Although you've now saved me from it, at some point I will descend to the depths of `beartype.door` myself. :D
